packages/repo-tools/src/commands/generate-catalog-info/generate-catalog-info.ts

```diff
@@ -0,0 +1,200 @@
+            await createCatalogInfoYaml({
+              yamlPath: infoYamlPath,
+              packageJson,
+              codeowners,
+              dryRun,
+            });
+            return;
+          }
+
+          throw e;
+        }
+
+        fixCatalogInfoYaml({
+          yamlPath: infoYamlPath,
+          packageJson,
+          codeowners,
+          yamlString,
+          dryRun,
+        });
+      }),
+    ),
+  );
+
+  const rejects = results.filter(
+    r => r.status === 'rejected',
+  ) as PromiseRejectedResult[];
+  if (rejects.length > 0) {
+    // Problems encountered. Print details here.
+    console.error(
+      chalk.red('Unable to create or fix catalog-info.yaml files\n'),
+    );
+    rejects.forEach(reject => console.error(`  ${reject.reason}`));
+    console.error();
+    process.exit(1);
+  }
+};
+
+type CreateOptions = {
+  yamlPath: string;
+  packageJson: BackstagePackageJson;
+  codeowners: CodeOwnersEntry[];
+  dryRun: boolean;
+};
+
+type FixOptions = CreateOptions & {
+  yamlString: string;
+};
+
+type BackstagePackageEntity = Entity & {
+  spec: {
+    type: 'backstage-package';
+    backstageRole: PackageRole;
+    lifecycle: string;
+    owner: string;
+    [key: string]: any;
+  };
+};
+
+function createCatalogInfoYaml(options: CreateOptions) {
+  const { codeowners, dryRun, packageJson, yamlPath } = options;
+  const owner = getOwnerFromCodeowners(codeowners, yamlPath);
+  const entity = createOrMergeEntity(packageJson, owner);
+
+  return dryRun
+    ? Promise.resolve(console.error(`Create ${relativePath('.', yamlPath)}`))
+    : writeFile(yamlPath, YAML.dump(entity));
+}
+
+function fixCatalogInfoYaml(options: FixOptions) {
+  const { codeowners, dryRun, packageJson, yamlPath, yamlString } = options;
+  const possibleOwners = getPossibleCodeowners(
+    codeowners,
+    relativePath('.', yamlPath),
+  );
+  const safeName = packageJson.name
+    .replace(/[^a-z0-9_\-\.]+/g, '-')
+    .replace(/^[^a-z0-9]|[^a-z0-9]$/g, '');
+  let yamlJson: BackstagePackageEntity;
+
+  try {
+    yamlJson = YAML.load(yamlString) as BackstagePackageEntity;
+  } catch (e) {
+    throw new Error(`Unable to parse ${relativePath('.', yamlPath)}: ${e}`);
+  }
+
+  const badOwner = !possibleOwners.includes(yamlJson.spec?.owner);
+  const badTitle = yamlJson.metadata.title !== packageJson.name;
+  const badName = yamlJson.metadata.name !== safeName;
+  const badType =
+    yamlJson.spec?.type !== `backstage-${packageJson.backstage.role}`;
+  const badDesc = yamlJson.metadata.description !== packageJson.description;
+
+  if (badOwner || badTitle || badName || badType || badDesc) {
+    const owner = badOwner
+      ? getOwnerFromCodeowners(codeowners, yamlPath)
+      : yamlJson.spec?.owner;
+    const newJson = createOrMergeEntity(packageJson, owner, yamlJson);
+    return dryRun
+      ? Promise.resolve(console.error(`Update ${relativePath('.', yamlPath)}`))
+      : writeFile(yamlPath, yamlOverwrite(yamlString, newJson));
```

Comment:
This thing is interesting 😁 curious how well it handles comments etc, gonna have to play around with it a bit